### ISSUE TITLE
feature-request: sks: export kubeconfig string (or even separate credentials) of the resource

### ISSUE THREAD

**Issue by Strangerxxx** [exoscale/terraform-provider-exoscale#131]
Hello!
I'd really like to have ability not just create SKS cluster using terraform, but at the same time get the connection credentials to use it in-place with kubernetes/helm terraform providers.
It is kind of enabler for me to use SKS on Exoscale.
As I can understand there is pretty straight forward possibility to implement such feature.
However if there is any cons which prevents of doing so please let me know.

PS: most of managed kubernetes providers like DO, GCP, AWS, YACloud have this ability.

**Comment by tstrohmeier** [comment]
+1

**Comment by retrack** [comment]
We had debated internally already before your issue and had concluded that it was not a safe practice to do. Happy to revisit however especially if other providers behave like so.

**Comment by secustor** [comment]
Examples of how other cloud providers handle this situation:
 
- Azure:
  directly exposes the kubeconfig as attributes. https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#kube_config_raw
- GCP: 
  exposes a datasource https://github.com/hashicorp/terraform-provider-kubernetes/blob/main/_examples/gke/main.tf#L26
- AWS: 
  does as Exoscale requires the usage of the kubeconfig exec plugin https://github.com/WesleyCharlesBlake/terraform-aws-eks/blob/081b9c703a4e23db0bbc9024c08e5cc62c7c08f4/config.tf#L37

**Comment by Strangerxxx** [comment]
> * AWS:
>   does as Exoscale requires the usage of the kubeconfig exec plugin

Could you explain how I may use exoscale cli as a kubeconfig exec plugin? @secustor

If I'm able to use it as plugin, than there is no reason to expose kubeconfig, unfortunately I did not found any documented example of this way usage of exoscale cli (

**Comment by secustor** [comment]
It eases the usage in CI systems as the user does not have to install and configure the `exo` cli separately to Terraform. 

This is the output of `exo compute sks kubeconfig --help`: 
```
    apiVersion: v1
    kind: Config
    clusters:
    - name: my-sks-cluster
      cluster:
        certificate-authority-data: **BASE64-ENCODED CLUSTER CERTIFICATE**
        server: https://153fcc53-1197-46ae-a8e0-ccf6d09efcb0.sks-ch-gva-2.exo.io:443
    users:
    - name: exo@my-sks-cluster
      user:
        # The "exec" section replaces "client-certificate-data"/"client-key-data"
        exec:
          apiVersion: "client.authentication.k8s.io/v1beta1"
          command: exo
          args:
          - sks
          - kubeconfig
          - my-sks-cluster
          - --zone=ch-gva-2
          - --exec-credential
          - user
    contexts:
    - name: my-sks-cluster
      context:
        cluster: my-sks-cluster
        user: exo@my-sks-cluster
    current-context: my-sks-cluster
```

**Comment by Strangerxxx** [comment]
awesome, that's what I'm looking for!

however the `certificate-authority-data` has to be exported of cluster resource in this case

**Comment by Strangerxxx** [comment]
@secustor could you please also explain why `certificate-authority-data` content of generated kubeconfig (using `exo compute sks kubeconfig`) doesn't match neither authorities retrieved using `exo compute sks authority-cert`?

I tried to authenticate using certs retrieved using `exo compute sks authority-cert` without success.

How I may retrieve cluster-ca using exo cli?

**Comment by secustor** [comment]
I can only guess which is the reason. Could be that this are different authorities of the cluster are returned, but this is question can be only answered by someone by Exoscale. 

I'm not sure how to do this best at the moment.   AWS returns the certificate authority as attribute and can therefore be templated into the kubeconfig. Exoscale does not expose this as attribute in the provider at the moment, but it can be requested using the API.        

When you retrieve the whole kubeconfig it has to contain the correct certificate authority.

**Comment by jessicatoscani** [comment]
@Strangerxxx The `certificate-authority-data content` of kubeconfig is the apiserver CA. The command `exo compute sks authority-cert` can only provide [kubelet or aggregation authorities](https://community.exoscale.com/documentation/sks/certificates-and-keys/#certificates-authorities)

```
exo compute sks authority-cert --help
This command retrieves the certificate content for the specified Kubernetes
cluster authority. Supported authorities:

Supported authorities: aggregation, kubelet
```
We have just released a new version of our provider with `exoscale_sks_kubeconfig` resource (#159 )

**Comment by fishman** [comment]
Various terraform state providers encrypt the state on rest. Gitlab does that for example.

Currently to be able to move secrets like the Kafka secrets into the cluster you first need to export them as a file or the env and then write them into the cluster.

If you provide exports for database and kubernetes secrets, database access can be written directly to kubernetes cluster in multistep terraform configs.

Terraform has a sensitive flag specifically for this purpose to make sure you really want to store the output in remote state:

```
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output containing sensitive data be
│ explicitly marked as sensitive, to confirm your intent.
│ 
│ If you do intend to export this data, annotate the output value as sensitive by adding the following argument:
│     sensitive = true
```

Right now you have to write wrappers wrappers around the exoscale CLI which is quite frankly nasty user experience.

**Comment by PhilippeChepy** [comment]
We can close this issue because as @jessicatoscani outlined, a solution was implemented in #159.

This update is available in our plugin since v0.33.0 [CHANGELOG](https://github.com/exoscale/terraform-provider-exoscale/blob/master/CHANGELOG.md#0330-march-11-2022). 

Example usages are available:
- in the [provider doc](https://registry.terraform.io/providers/exoscale/exoscale/latest/docs/resources/sks_kubeconfig), and
-  in [this repository](https://github.com/exoscale/terraform-provider-exoscale/tree/master/examples/sks)
